### ISSUE TITLE
Avoid indexing ctypes.WinError objects

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5364]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5364|
|Type|defect|
|Created|2011-11-04 20:45:01Z|

Attaching patch, which replaces indexing with a lookup of the ``strerror`` attribute.

Attachments:

* [winerror.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/857/857e17ffc4781a2ec016adfe176349d1719eb34a/bf6940410bb27f05f78d60851655198d6529efe4.patch) (485 bytes) - added by antoine on 2011-11-04 20:45:21Z - 
* [winerror2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/857/857e17ffc4781a2ec016adfe176349d1719eb34a/83c957e3dbbc822d3a69c04b7e5039fc5048decd.patch) (1207 bytes) - added by antoine on 2011-11-06 15:07:06Z - 
* [winerror3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/857/857e17ffc4781a2ec016adfe176349d1719eb34a/d0987f746f02040c8a61edd1dfebf96c12717d17.patch) (1870 bytes) - added by antoine on 2011-11-06 17:51:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5364 5364
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__Python_3_x Python-3.x
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1320439501000000 1320439501000000
changetime__1320641954000000 1320641954000000
version__None None
owner__exarkun exarkun
cc__antoine
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @pitrou|
|-|-|

Thanks.  Took me a few minutes to find the unit tests for this code, which turned out to be in `twisted.test.test_strerror` instead of `twisted.python.test.test_win32`.  We should probably merge those two test modules.

The change seems somewhat sensible, but I get a test failure:

```

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_strerror.py", line 89, in test_winError
    message = formatter.formatError(self.probeErrorCode)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/win32.py", line 159, in formatError
    return self.winError(errorcode).strerror
exceptions.AttributeError: 'tuple' object has no attribute 'strerror'

twisted.test.test_strerror.ErrorFormatingTestCase.test_winError
-------------------------------------------------------------------------------
```

(on either Linux or Windows).  Looks like the tests need to be adjusted a bit as well.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

I am sorry, I hadn't run the test suite. Here is an updated patch fixing the tests (since I think passing a tuple here is bogus). Tested under Windows.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)|@pitrou removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_7"></a>|@pitrou commented|
|-|-|

New patch fixing additional instances of exception indexing in test_strerror itself (sorry, I was initially testing under 2.x).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This looks good and passes tests here as well.  I'll apply it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [33091]) Use the strerror attribute of ctypes.WinError instead of indexing the exception object.

Author: antoine
Reviewer: exarkun
Fixes: [#5364](https://github.com/twisted/twisted/issues/5364)

Replace another usage of the deprecated exception indexing API with a named
attribute of the exception object.  In this case, the `strerror` attribute of
the `ctypes.WinError` type.
